---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# Latent Embedding Multivariate Regression (LEMUR)

<!-- badges: start -->
<!-- badges: end -->

The goal of lemur is to enable easy analysis of multi-condition single-cell data. Lemur fits latent embedding regression, which means that it tries to find the PCA embedding for each condition and parameterizes the transition from one embedding to the other. For this task, lemur uses geodesic regression on the Grassmann manifold, which is solved efficiently using tangent-space linear modelling. The result is an interpretable model of the gene expression for arbitrary experimental designs that can be expressed using a design matrix.

## Installation

You can install the released version of lemur from Github

```r
devtools::install_github("const-ae/lemur")
```

## Example

Make some data

```{r generate-data}
mat <- matrix(rnorm(30 * 500), nrow = 30, ncol = 500)
col_data <- data.frame(condition = sample(letters[1:2], 500, replace = TRUE))
```

Fit the model

```{r fit-differential-embedding-model}
fit <- lemur::lemur(mat, design = ~ condition, col_data = col_data, n_ambient = 10, n_embedding = 2)
fit
```

Let's look at the coefficients

```{r show-fit-results}
round(fit$coefficients, 5)
plot(t(fit$embedding), col = as.factor(col_data$condition))
```


```{r}
de_mat <- lemur::test_de(fit, contrast = cond(condition = "a") == cond(condition = "b"))
de_mat[1:5, 1:10]
```

Show the gene expression changes on the latent embedding
```{r visualize-differential-expression}
pal <- scales::col_numeric(scales::viridis_pal()(50), NULL)
plot(t(fit$embedding), col = pal(de_mat[10,]), pch = 16, cex = 1.3)
```


```{r}
counts <- matrix(rpois(30 * 500, lambda = 2.4), nrow = 30, ncol = 500)
SummarizedExperiment::colData(fit)$patient_id <- sample(c("A", "B", "C"), size = 500, replace = TRUE)
de_regions <- lemur::find_de_neighborhoods(fit, de_mat, counts = counts, group_by = glmGamPoi::vars(patient_id, condition), 
                             contrast = cond(condition = "a") == cond(condition = "b"))
head(de_regions)
```


# Session Info

```{r}
sessionInfo()
```


