---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# DiffEmbSeq

<!-- badges: start -->
<!-- badges: end -->

The goal of DiffEmbSeq is to enable easy analysis of multi-condition single-cell data. DiffEmbSeq fits a differential embedding model, which means that it tries to find the PCA embedding for each condition and parameterizes the transition from one embedding to the other. For this task, DiffEmbSeq uses geodesic regression on the Grassmann manifold, which is solved efficiently using tangent-space linear modelling. The result is an interpretable model of the gene expression for arbitrary experimental designs that can be expressed using a design matrix.

## Installation

You can install the released version of DiffEmbSeq from https://git.embl.de/ahlmanne/DiffEmbSeq

To get read access you have set the following _access token_ which is valid until the 31.08.2022.
``` r
Sys.setenv(GITLAB_PAT = "glpat-_2uHibDPXsgcs9kYxMsM")
remotes::install_gitlab(repo = "ahlmanne/DiffEmbSeq", host = "https://git.embl.de/")
```

## Example

Make some data

```{r generate-data}
mat <- matrix(rnorm(30 * 500), nrow = 30, ncol = 500)
col_data <- data.frame(condition = sample(letters[1:2], 500, replace = TRUE))
```

Fit the model

```{r fit-differential-embedding-model}
fit <- DiffEmbSeq::differential_embedding(mat, design = ~ condition, col_data = col_data,
                                          n_ambient = 10, n_embedding = 2)
fit
```

Let's look at the coefficients

```{r show-fit-results}
round(fit$diffemb_coefficients, 5)
plot(t(fit$diffemb_embedding), col = as.factor(col_data$condition))
```

Bootstrap to get an estimate of the parameter variance

```{r bootstrap-variance}
fit <- DiffEmbSeq::estimate_variance(fit, n_bootstrap_samples = 30)
```


```{r}
res <- DiffEmbSeq::test_differential_expression(fit, contrast = fact(condition = "a") == fact(condition = "b"),
                                                consider = "embedding+linear", variance_est = "bootstrap", 
                                                return = "table")
head(res)
```

Show the gene expression changes on the latent embedding
```{r visualize-differential-expression}
pal <- scales::col_numeric(scales::viridis_pal()(50), NULL)
plot(t(fit$diffemb_embedding), col = pal(res[res$feature == "feature_10",]$diff),
     pch = 16, cex = 1.3)
```




